compiler/src/iree/compiler/DispatchCreation/SetEncoding.cpp

```diff
@@ -227,17 +229,45 @@ class SetContractionOpEncoding final
     SmallVector<AffineMap> maps = linalgOp.getIndexingMapsArray();
 
     auto opType = IREE::Encoding::EncodingOpType::matmul;
-    auto setEncodingWrapper = [&](Value src, int64_t operandIndex) -> Value {
-      auto encoding =
-          EncodingAttr::get(linalgOp.getContext(), operandIndex, opType,
-                            elemTypes, maps, iterationSizes);
+    auto setEncodingWrapper = [&](Value src,
+                                  int64_t operandIndex) -> FailureOr<Value> {
+      MLIRContext *ctx = linalgOp.getContext();
+      Attribute encoding;
+      switch (encodingOption) {
+      case EncodingOptions::Default: {
+        encoding = EncodingAttr::get(ctx, operandIndex, opType, elemTypes, maps,
+                                     iterationSizes);
+        break;
+      }
+      case EncodingOptions::MatmulK: {
+        SmallVector<int32_t> kDims;
+        AffineMap indexingMap = maps[operandIndex];
+        auto cDims = linalg::inferContractionDims(linalgOp);
+        for (auto k : cDims->k) {
+          std::optional<unsigned> dimIdx =
+              indexingMap.getResultPosition(rewriter.getAffineDimExpr(k));
+          if (!dimIdx) {
+            continue;
+          }
+          kDims.push_back(dimIdx.value());
+        }
+        encoding = MatmulKAttr::get(ctx, rewriter.getDenseI32ArrayAttr(kDims));
+        break;
+      }
+      default: {
+        return failure();
+      }
+      }
       return setEncoding(rewriter, loc, src, encoding);
     };
-    Value encodedLhs = setEncodingWrapper(lhs, IREE::Encoding::MATMUL_LHS);
-    Value encodedRhs = setEncodingWrapper(rhs, IREE::Encoding::MATMUL_RHS);
-    Value encodedOut = setEncodingWrapper(out, IREE::Encoding::MATMUL_RESULT);
-    Value opTiled = clone(rewriter, linalgOp, encodedOut.getType(),
-                          ValueRange{encodedLhs, encodedRhs, encodedOut})
+    auto encodedLhs = setEncodingWrapper(lhs, IREE::Encoding::MATMUL_LHS);
+    auto encodedRhs = setEncodingWrapper(rhs, IREE::Encoding::MATMUL_RHS);
+    auto encodedOut = setEncodingWrapper(out, IREE::Encoding::MATMUL_RESULT);
+    if (failed(encodedLhs) || failed(encodedRhs) || failed(encodedOut)) {
+      return failure();
+    }
+    Value opTiled = clone(rewriter, linalgOp, encodedOut->getType(),
+                          ValueRange{*encodedLhs, *encodedRhs, *encodedOut})
```

Comment:
We can revert the change if you follow the other comment: return `Value()` instead.